pkg/roachprod/prometheus/prometheus.go

```diff
@@ -0,0 +1,480 @@
+	if len(nodes) != len(ips) {
+		return errors.New("number of nodes must exactly match the number of ips")
+	}
+	sn := ScrapeNode{Nodes: nodes, IPs: ips, Port: port}
+	for i := range cfg.ScrapeConfigs {
+		sc := &cfg.ScrapeConfigs[i]
+		// A workload scrape config name is unique
+		if sc.JobName == name {
+			sc.ScrapeNodes = append(sc.ScrapeNodes, sn)
+			return nil
+		}
+	}
+	cfg.ScrapeConfigs = append(cfg.ScrapeConfigs, MakeWorkloadScrapeConfig(name, []ScrapeNode{sn}))
+	return nil
+}
+
+// WithPrometheusNode specifies the nodes to set up prometheus on.
+func (cfg *Config) WithPrometheusNode(node install.Node) *Config {
+	cfg.PrometheusNode = install.Nodes{node}
+	return cfg
+}
+
+// WithCluster adds scraping for a CockroachDB cluster running on the given nodes.
+// Chains for convenience.
+func (cfg *Config) WithCluster(nodes install.Nodes, ips []string) error {
+	if len(nodes) != len(ips) {
+		return errors.New("number of nodes must exactly match the number of ips")
+	}
+	cfg.ScrapeConfigs = append(cfg.ScrapeConfigs, MakeInsecureCockroachScrapeConfig(nodes, ips)...)
+	return nil
+}
+
+// WithGrafanaDashboard adds links to dashboards to provision into Grafana. See
+// cfg.Grafana.DashboardURLs for helpful tips.
+// Enables Grafana if not already enabled.
+// Chains for convenience.
+func (cfg *Config) WithGrafanaDashboard(url string) *Config {
+	cfg.Grafana.Enabled = true
+	cfg.Grafana.DashboardURLs = append(cfg.Grafana.DashboardURLs, url)
+	return cfg
+}
+
+// WithNodeExporter causes node_exporter to be set up on the specified machines,
+// a separate process that sends hardware metrics to prometheus.
+func (cfg *Config) WithNodeExporter(nodes install.Nodes, ips []string) error {
+	if len(nodes) != len(ips) {
+		return errors.New("number of nodes must exactly match the number of ips")
+	}
+	cfg.NodeExporter = nodes
+	cfg.NodeExporterIPs = ips
+	return nil
+}
+
+// Prometheus contains metadata of a running instance of prometheus.
+type Prometheus struct {
+	Config
+}
+
+// Init creates a prometheus instance on the given cluster.
+func Init(
+	ctx context.Context, l *logger.Logger, c *install.SyncedCluster, cfg Config,
+) (_ *Prometheus, _ error) {
+	if len(cfg.NodeExporter) > 0 {
+		if err := c.RepeatRun(ctx, l, os.Stdout, os.Stderr, cfg.NodeExporter,
+			"download node exporter",
+			`
+(sudo systemctl stop node_exporter || true) &&
+rm -rf node_exporter && mkdir -p node_exporter && curl -fsSL \
+  https://github.com/prometheus/node_exporter/releases/download/v1.3.1/node_exporter-1.3.1.linux-amd64.tar.gz |
+  tar zxv --strip-components 1 -C node_exporter
+`); err != nil {
+			return nil, err
+		}
+
+		// Start node_exporter.
+		if err := c.Run(ctx, l, os.Stdout, os.Stderr, cfg.NodeExporter, "init node exporter",
+			`cd node_exporter &&
+sudo systemd-run --unit node_exporter --same-dir ./node_exporter`,
+		); err != nil {
+			return nil, err
+		}
+
+		// Add a scrape config for each node exporter
+		for i, node := range cfg.NodeExporter {
```

Comment:
why not have a config helper function for nodeExporter as well? Is there any reason this needs to be during during init?

Reply:
No, I think I was just being a little shoddy here, I added the config helpers later but guess I didn't update this one